In [4]:
import agate

In [5]:
unlcrimes = agate.Table.from_csv ('Data/unlcrime.csv')

In [6]:
unlcrimes.print_table()

| Case #     | Incident Code        |            Reported | Case Status          |      Start Occurred |        End Occurred | ... |
| ---------- | -------------------- | ------------------- | -------------------- | ------------------- | ------------------- | --- |
| 10004605.0 | VANDALISM - BY GR... | 2010-11-19 05:55:00 | Inactive             | 2010-11-18 12:00:00 | 2010-11-19 05:00:00 | ... |
| 10004607.0 | NARCOTICS - POSSE... | 2010-11-19 09:01:00 | Cleared by Arrest    | 2010-11-19 09:01:00 |                     | ... |
| 10004608.0 | VANDALISM - BY GR... | 2010-11-19 09:32:00 | Inactive             | 2010-11-18 11:00:00 | 2010-11-19 09:32:00 | ... |
| 10004609.0 | VANDALISM - OTHER    | 2010-11-19 10:53:00 | Inactive             | 2010-11-18 22:00:00 | 2010-11-19 10:53:00 | ... |
| 10004610.0 | LITTERING            | 2010-11-19 12:02:00 | Inactive             | 2010-11-16 17:00:00 | 2010-11-19 12:02:00 | ... |
| 10004614.0 | LOST OR STOLEN ITEM  | 2010-11-19 16:17:00 | Inactive 

In [7]:
print(unlcrimes)

| column         | data_type |
| -------------- | --------- |
| Case #         | Text      |
| Incident Code  | Text      |
| Reported       | DateTime  |
| Case Status    | Text      |
| Start Occurred | DateTime  |
| End Occurred   | DateTime  |
| Building       | Text      |
| Location       | Text      |
| Stolen         | Number    |
| Damaged        | Number    |
| Description    | Text      |



In [8]:
codes = unlcrimes.group_by('Incident Code')
code_counts = codes.aggregate([
    ('count', agate.Count())
])
code_counts = code_counts.order_by('Incident Code', reverse=True)

In [9]:
code_counts.print_table()

| Incident Code        | count |
| -------------------- | ----- |
|                      |     1 |
| WEAPONS – OTHER    |    10 |
| WEAPONS - ILLEGAL... |     1 |
| WEAPONS - FELON I... |     3 |
| WEAPONS - DISCHAR... |     3 |
| WEAPONS - CONCEALED  |     7 |
| WARRANT SERVICE      |     4 |
| VANDALISM - OTHER    |   392 |
| VANDALISM - BY GR... |   106 |
| UNL POLICY VIOLAT... |    30 |
| UNL POLICY VIOLAT... |    12 |
| TRESPASSING          |   311 |
| TRAFFIC - SUSPEND... |   707 |
| TRAFFIC - OTHER      |    54 |
| TRAFFIC - IMPROPE... |     2 |
| TRAFFIC - DIRECTION  |     1 |
| TELEPHONE - THREA... |    15 |
| TELEPHONE - OTHER    |    56 |
| TELEPHONE - OBSCE... |     6 |
| SUSPICIOUS VEHICLE   |     1 |
| ...                  |   ... |


In [10]:
crimes_with_years = unlcrimes.compute([
    ('reported_year', agate.Formula(agate.Text(), lambda row: '%s' % row['Reported'].year))
])

In [11]:
years = crimes_with_years.group_by('reported_year')
year_counts = years.aggregate([
    ('count', agate.Count())
])
year_counts.print_table()

| reported_year | count |
| ------------- | ----- |
| 2010          |   185 |
| 2011          | 1,948 |
| 2012          | 1,881 |
| 2013          | 2,052 |
| 2014          | 1,920 |
| 2015          | 1,750 |


In [12]:
crimes_with_monthyears = unlcrimes.compute([
    ('reported_monthyear', agate.Formula(agate.Text(), lambda row: '%s/%s' % (row['Reported'].month, row['Reported'].year)))
])

In [13]:
monthyears = crimes_with_monthyears.group_by('reported_monthyear')
monthyear_counts = monthyears.aggregate([
    ('count', agate.Count())
])
monthyear_counts.print_bars('reported_monthyear', 'count', width=60)

reported_monthyear count
11/2010               56 ▓░░░░░░                            
12/2010              129 ▓░░░░░░░░░░░░░░░                   
1/2011               157 ▓░░░░░░░░░░░░░░░░░░                
2/2011               168 ▓░░░░░░░░░░░░░░░░░░░               
3/2011               157 ▓░░░░░░░░░░░░░░░░░░                
4/2011               184 ▓░░░░░░░░░░░░░░░░░░░░░             
5/2011               130 ▓░░░░░░░░░░░░░░░                   
6/2011               124 ▓░░░░░░░░░░░░░░                    
7/2011               137 ▓░░░░░░░░░░░░░░░░                  
8/2011               183 ▓░░░░░░░░░░░░░░░░░░░░░             
9/2011               242 ▓░░░░░░░░░░░░░░░░░░░░░░░░░░░       
10/2011              175 ▓░░░░░░░░░░░░░░░░░░░░              
11/2011              179 ▓░░░░░░░░░░░░░░░░░░░░              
12/2011              112 ▓░░░░░░░░░░░░░                     
1/2012               194 ▓░░░░░░░░░░░░░░░░░░░░░░            
2/2012               147 ▓░░░░░░░░░░░░░░░░░                 

    ***MISSING DATA LOCATION***

In [14]:
codes = unlcrimes.group_by('Location')
code_counts = codes.aggregate([
    ('count', agate.Count())
])
code_counts = code_counts.order_by('Location', reverse=True)

In [15]:
code_counts.print_table()

| Location             | count |
| -------------------- | ----- |
| Y ST & N 33RD ST ... |     1 |
| Y ST & N 27TH ST ... |     4 |
| Y ST & N 23RD ST ... |     1 |
| Y ST & N 22ND ST ... |     1 |
| Y ST & N 21ST ST ... |     2 |
| Y ST & N 21ST ST ... |     2 |
| Y ST & N 19TH ST ... |     4 |
| Y ST & N 19TH ST ... |     1 |
| Y ST & N 19TH ST ... |     1 |
| Y ST & N 17TH ST ... |     7 |
| X ST & N 17TH ST ... |     1 |
| X ST & N 17TH ST ... |     1 |
| X ST & N 16TH ST ... |     5 |
| X ST & N 16TH ST ... |     1 |
| WHITTIER ST & W S... |     1 |
| W ST & N 25TH ST ... |     1 |
| W ST & N 23RD ST ... |     1 |
| W ST & N 22ND ST ... |     1 |
| W ST & N 16TH ST ... |     1 |
| W ST & N 16TH ST ... |     1 |
| ...                  |   ... |


***MISSING DATA BUILDING***

In [16]:
codes = unlcrimes.group_by('Building')
code_counts = codes.aggregate([
    ('count', agate.Count())
])
code_counts = code_counts.order_by('Building', reverse=True)

In [17]:
code_counts.print_table()

| Building             | count |
| -------------------- | ----- |
|                      | 3,506 |
| Wick Alumni Center   |    15 |
| Whittier Building    |    31 |
| Westbrook Music B... |    37 |
| West Stadium         |     4 |
| Wells Fargo Building |     1 |
| Water Sciences Lab   |     1 |
| Warehouse 2          |     2 |
| Warehouse 1          |     2 |
| Walter Scott Engi... |    21 |
| Vine Street Rec F... |     7 |
| Veterinary Diagno... |     1 |
| Vet Basic Sciences   |     3 |
| Varner Hall          |     8 |
| Various              |     1 |
| Utility Plant, East  |     2 |
| Utility Plant, City  |     8 |
| Unknown              |     1 |
| University Suites    |    75 |
| University Park A... |     5 |
| ...                  |   ... |


**DETECTED MISSING DATA IN BUILDING 3,506**

**

***STOLEN***

In [18]:
codes = unlcrimes.group_by('Stolen')
code_counts = codes.aggregate([
    ('count', agate.Count())
])
code_counts = code_counts.order_by('Stolen', reverse=True)

In [19]:
code_counts.print_table()

|    Stolen | count |
| --------- | ----- |
|           | 8,260 |
| 52,965.00 |     1 |
| 50,053.00 |     1 |
| 23,337.00 |     1 |
| 22,000.00 |     1 |
| 20,348.00 |     1 |
| 15,480.00 |     1 |
| 15,000.00 |     1 |
| 14,700.00 |     1 |
| 13,000.00 |     1 |
|  9,954.08 |     1 |
|  9,500.00 |     1 |
|  8,800.00 |     1 |
|  8,300.00 |     1 |
|  8,000.00 |     1 |
|  7,505.00 |     1 |
|  7,300.00 |     1 |
|  7,205.00 |     1 |
|  7,000.00 |     1 |
|  6,000.00 |     2 |
|       ... |   ... |


**DETECTED MISSING DATA IN STOLEN: 8,260**

***DAMAGED***

In [20]:
codes = unlcrimes.group_by('Damaged')
code_counts = codes.aggregate([
    ('count', agate.Count())
])
code_counts = code_counts.order_by('Damaged', reverse=True)

In [21]:
code_counts.print_table()

| Damaged | count |
| ------- | ----- |
|         | 8,455 |
| 100,000 |     1 |
|  50,000 |     1 |
|  14,000 |     1 |
|  10,000 |     1 |
|   8,800 |     1 |
|   8,500 |     1 |
|   8,000 |     1 |
|   7,700 |     1 |
|   6,500 |     1 |
|   6,180 |     1 |
|   6,000 |     2 |
|   5,600 |     1 |
|   5,500 |     1 |
|   5,000 |     4 |
|   4,850 |     1 |
|   4,700 |     1 |
|   4,500 |     4 |
|   4,300 |     1 |
|   4,000 |     5 |
|     ... |   ... |


**DETECTED MISSING DATA IN DAMAGED: 8,455**

In [22]:
codes = unlcrimes.group_by('Start Occurred')
code_counts = codes.aggregate([
    ('count', agate.Count())
])
code_counts = code_counts.order_by('Start Occurred', reverse=True)

In [23]:
code_counts.print_table()

|      Start Occurred | count |
| ------------------- | ----- |
|                     |   475 |
| 2015-12-22 14:00:00 |     1 |
| 2015-11-18 22:28:00 |     1 |
| 2015-11-18 21:52:00 |     1 |
| 2015-11-18 21:08:00 |     1 |
| 2015-11-18 13:23:00 |     1 |
| 2015-11-18 01:14:00 |     1 |
| 2015-11-17 02:54:00 |     1 |
| 2015-11-17 00:36:00 |     1 |
| 2015-11-17 00:18:00 |     1 |
| 2015-11-16 23:13:00 |     1 |
| 2015-11-16 22:19:00 |     1 |
| 2015-11-16 15:30:00 |     1 |
| 2015-11-16 01:32:00 |     1 |
| 2015-11-15 19:10:00 |     1 |
| 2015-11-15 18:15:00 |     1 |
| 2015-11-15 06:34:00 |     1 |
| 2015-11-15 02:42:00 |     1 |
| 2015-11-15 02:35:00 |     1 |
| 2015-11-15 00:47:00 |     1 |
|                 ... |   ... |


***DETECTED MISSING START OCCURED DATA: 475***

In [24]:
codes = unlcrimes.group_by('End Occurred')
code_counts = codes.aggregate([
    ('count', agate.Count())
])
code_counts = code_counts.order_by('End Occurred', reverse=True)

In [25]:
code_counts.print_table()

|        End Occurred | count |
| ------------------- | ----- |
|                     | 6,830 |
| 2015-11-18 13:24:00 |     1 |
| 2015-11-17 10:48:00 |     1 |
| 2015-11-17 10:15:00 |     1 |
| 2015-11-16 15:15:00 |     1 |
| 2015-11-16 13:12:00 |     1 |
| 2015-11-16 11:30:00 |     1 |
| 2015-11-16 06:00:00 |     1 |
| 2015-11-15 11:00:00 |     1 |
| 2015-11-15 02:04:00 |     1 |
| 2015-11-14 17:37:00 |     1 |
| 2015-11-12 10:00:00 |     1 |
| 2015-11-11 16:22:00 |     1 |
| 2015-11-11 08:00:00 |     1 |
| 2015-11-10 09:15:00 |     1 |
| 2015-11-10 08:00:00 |     1 |
| 2015-11-08 19:46:00 |     1 |
| 2015-11-07 17:00:00 |     1 |
| 2015-11-06 17:42:00 |     1 |
| 2015-11-04 07:34:00 |     1 |
|                 ... |   ... |


***DETECTED MISSING END OCCURRED DATA: 6,830***

In [47]:
print(unlcrimes)

| column         | data_type |
| -------------- | --------- |
| Case #         | Text      |
| Incident Code  | Text      |
| Reported       | DateTime  |
| Case Status    | Text      |
| Start Occurred | DateTime  |
| End Occurred   | DateTime  |
| Building       | Text      |
| Location       | Text      |
| Stolen         | Number    |
| Damaged        | Number    |
| Description    | Text      |



In [50]:
print (len(unlcrimes.rows))

9736


In [51]:
by_status = unlcrimes.group_by('Incident Code','Case Status')

In [54]:
print(by_status)

| table                               | rows |
| ----------------------------------- | ---- |
| VANDALISM - BY GRAFFITI             | 106  |
| NARCOTICS - POSSESSION              | 607  |
| VANDALISM - OTHER                   | 392  |
| LITTERING                           | 161  |
| LOST OR STOLEN ITEM                 | 425  |
| ALCOHOL - MINOR IN POSSESSION       | 954  |
| ALCOHOL - DRUNK                     | 543  |
| ACCIDENTS - P.D. NOT REPORTABLE     | 213  |
| SS - CHECK WELFARE OF PERSON        | 142  |
| LARCENY - FROM MOTOR VEHICLE        | 135  |
| LARCENY - FROM BUILDING             | 370  |
| ACCIDENTS - P.D. H&R NOT REPORTABLE | 327  |
| MEDICAL EMERGENCY                   | 895  |
| LARCENY - STOLEN BIKE               | 464  |
| ACCIDENTS - P.D. H&R REPORTABLE     | 15   |
| TRESPASSING                         | 311  |
| ALCOHOL - DWI                       | 558  |
| ASSAULT - DOMESTIC                  | 18   |
| ASSAULT - NON DOMESTIC              | 60   |
| FIRE (WORKI

In [55]:
print (len(by_status))

117


In [56]:
status_totals = by_status.aggregate ([
    ('count', agate.Count())
])

In [57]:
print(status_totals)

| column      | data_type |
| ----------- | --------- |
| Case Status | Text      |
| count       | Number    |



In [58]:
sorted_status = status_totals.order_by('count', reverse =True)

In [59]:
sorted_status.print_table()

| Case Status          | count |
| -------------------- | ----- |
| ALCOHOL - MINOR I... |   954 |
| MEDICAL EMERGENCY    |   895 |
| TRAFFIC - SUSPEND... |   707 |
| NARCOTICS - POSSE... |   607 |
| ALCOHOL - DWI        |   558 |
| ALCOHOL - DRUNK      |   543 |
| DISTURBANCE - OTHER  |   467 |
| LARCENY - STOLEN ... |   464 |
| LOST OR STOLEN ITEM  |   425 |
| VANDALISM - OTHER    |   392 |
| LARCENY - FROM BU... |   370 |
| ACCIDENTS - P.D. ... |   327 |
| TRESPASSING          |   311 |
| FIRE - FALSE ALARM   |   215 |
| ACCIDENTS - P.D. ... |   213 |
| LITTERING            |   161 |
| PROPERTY DAMAGE -... |   152 |
| LARCENY - OTHER O... |   145 |
| SS - CHECK WELFAR... |   142 |
| OUTSIDE - O.P.S. ... |   136 |
| ...                  |   ... |


*** BASED ON THE DATA OF CRIMES HAPPENED AT THE UNL, WE FIGURED OUT THAT ALCOHOL RANKS THE FIRST PLACE OF TOP TEN CRIMES HAT THE UNL. TOP 10 CRIMES AT THE UNL: ALCOHOL - MINOR I (954 CASES), MEDICAL EMERGENCY (895 CASES), TRAFFIC SUSPEND (707 CASES), NARCOTICS POSSESIONS (607), ALCOHOL DWI (558 CASES), ALCOHOL DRUNK (543 CASES), DISTURBANCE (467 CASES), LARCENY -STOLEN (464 CASES), LOST OR STOLEN (425 CASES), AND VANDALISM (392 CASES)***

DATA READING:
1. If I get this kind of data from the government I will ask the more complete one since many rows are left blank. Is it because it is error? incomplete process of transfering data? or is there any something that is hidden by the government?
2. On column CASE STATUS, there are lots of explanation of inactive case. Why is it inactive? what does it mean? And what is the difference between cleared by arrest and cleared by exception, etc?
3. I will ask to the police department since it deals with crime, why it has not been completed? can they provide the more completed ones? 
4. OUTLIERS refers to values that are extremely high or extremely low. Therefore, on the unlcrimes data, we can find outliers from the reported year. E.g. total reported crimes in 2010 shows only 185 cases, while the next year data shows 1,948 cases. So, value of crimes in 2010 is considered extremely low. It is an outlier. Further analysis should be made out of it.